students/mayc4t/lesson09/mailroom5.py

```diff
@@ -0,0 +1,239 @@
+        str_ret = "\t{:<30}|{:>14} | {:>12} | {:>14}".format(
+            self.name, self.total_donation, self.num_gift, self.avg)
+        return (str_ret)
+
+    def __repr__(self):
+        return "'({} {} {})'".format(self.name, self._num_gift, self._avg)
+
+    def print_thank(self, amt):
+        print("\n\n\t\tDear {},\n\t\tThank you for your recent generous donation of {}.\n\t\tThanks,".format(
+            self.name, amt))
+
+    def add_donation(self, amount):
+        self._donations.append(amount)
+
+
+class Donor_DB():
+    # You’ll then want a class that handles the collection of donors.
+    # This will hold all the donor objects,
+    #       but also methods to add a new donor,
+    #       search for a given donor, etc.
+    #       If you want a way to save and re-load your data, this class would have that too.
+    # Your class for the collection of donors will also hold the code
+    # that generates reports about multiple donors.
+    def __init__(self, donors=None):
+        if donors is None:
+            self.donors_db = []
+        else:
+            self._donors_db = donors
+
+    def print_donors_name_list(self):
+        print ("\n\tList of Donor Name")
+        print ("\t------------------")
+        for dn_idx in self._donors_db:
+            print("\t\t{}".format(dn_idx.name))
+
+    @property
+    def names(self):
+        name_list = []
+        for dn_idx in self._donors_db:
+            name_list.append(dn_idx.name)
+        self._names = name_list
+        return self._names
+
+    def create_report(self):
+        sorted_list = sorted(
+            self._donors_db, key=Donor.sort_by_total, reverse=True)
+        print ("\n\n\tDonor.DB.create_report")
+        print("\t{:<30}|{:>14} | {:>12} | {:>14}".format(
+            "Donor Name", "Total Gift", "Num Gifts", "Avg Gift"))
+        print("\t" + "_"*30 + " " + "_"*14 + "   " + "_"*12 + "   " + "_"*14)
+        for dn_idx in sorted_list:
+            print(dn_idx.__str__())
+
+    def send_letters(self):
+        for dn in self._donors_db:
+            name = dn.name.split()
+            fname = name[0]+name[1] + ".txt"
+            print (fname)
+            try:
+                with open(fname, 'w') as outfile:
+                    outfile.write(
+                        "Dear {},\n\nThank you for your total generous donations: ${}.\n"
+                        "This will be put to good use. \n\nThanks".format(dn.name, dn.total_donation))
+            except PermissionError:
+                print ("Permission denied, can't open file {}".format(fname))
+
+    def add_donation(self, donor):
+        # check if donor is in self._donors_db
+        if any(donor.name in x for x in self.names):
+            idx = self.names.index(donor.name)
+            #print ("Exist")
+            #print (idx)
+            gift = donor.total_donation
+            self._donors_db[idx].add_donation(gift)
+            #print (self._donors_db[idx].__str__())
```

Comment:
It's a good idea to delete commented out debugging statements like this before committing